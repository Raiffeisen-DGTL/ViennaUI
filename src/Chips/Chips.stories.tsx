import React from 'react';
import { Story, Meta } from 'storybook';
import { Chips, ChipsProps } from './Chips';
import { Tooltip } from '../Tooltip';

export default {
    title: 'Development/Chips',
    component: Chips,
} as Meta;

export const Overview: Story<ChipsProps> = (args) => {
    const [active, setActive] = React.useState();
    const onClick = (e) => setActive(e.target.id);

    return (
        <Chips {...args} active={active} onClick={onClick} {...args}>
            <Chips.Item size='l' id='1' tabIndex={1}>
                Chip 1
            </Chips.Item>
            <Chips.Item size='l' id='2' tabIndex={1}>
                Chip 2
            </Chips.Item>
            <Chips.Item size='l' id='3' tabIndex={1}>
                Chip 3
            </Chips.Item>
            <Chips.Item size='l' id='4' tabIndex={1} disabled>
                Chip 3
            </Chips.Item>
        </Chips>
    );
};

export const MultipleSelection: Story<ChipsProps> = (args) => {
    const [active, setActive] = React.useState<string[]>([]);
    const onClick = (event) => {
        const id = event.target.id;
        setActive((state) => (state.includes(id) ? state.filter((i) => i !== id) : [...state, id]));
    };

    return (
        <Chips active={active} onClick={onClick} {...args}>
            <Chips.Item id='1' tabIndex={1} role='checkbox'>
                Chip 1
            </Chips.Item>
            <Chips.Item id='2' tabIndex={1} role='checkbox'>
                Chip 2
            </Chips.Item>
            <Chips.Item id='3' tabIndex={1} role='checkbox'>
                Chip 3
            </Chips.Item>
            <Chips.Item id='4' tabIndex={1} role='checkbox' disabled>
                Chip 4
            </Chips.Item>
        </Chips>
    );
};
MultipleSelection.storyName = 'Выбрать несколько';

export const ChipsWithTooltip: Story<ChipsProps> = (args) => {
    const [active, setActive] = React.useState();
    const onClick = (e) => setActive(e.currentTarget.id);

    return (
        <Chips {...args} active={active} onClick={onClick} {...args}>
            <Tooltip action={'hover'} content={'Chip 1 Tooltip text'}>
                <Chips.Item size='l' id='1' tabIndex={1}>
                    Chip 1
                </Chips.Item>
            </Tooltip>
            <Tooltip action={'hover'} content={'Chip 2 Tooltip text'}>
                <Chips.Item size='l' id='2' tabIndex={1}>
                    Chip 2
                </Chips.Item>
            </Tooltip>
            <Tooltip action={'hover'} content={'Chip 3 Tooltip text'}>
                <Chips.Item size='l' id='3' tabIndex={1}>
                    Chip 3
                </Chips.Item>
            </Tooltip>
            <Tooltip action={'hover'} content={'Chip 4 Tooltip text'}>
                <Chips.Item size='l' id='4' tabIndex={1} disabled>
                    Chip 4
                </Chips.Item>
            </Tooltip>
        </Chips>
    );
};
ChipsWithTooltip.storyName = 'Chips с Tooltip внутри';

export const AllDesigns: Story<ChipsProps> = (args) => {
    return (
        <>
            <Chips design='accent'>
                <Chips.Item id='1' tabIndex={1} active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' tabIndex={1}>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' tabIndex={1}>
                    Chip
                </Chips.Item>
            </Chips>
            <Chips design='primary'>
                <Chips.Item id='1' tabIndex={1} active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' tabIndex={1}>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' tabIndex={1}>
                    Chip
                </Chips.Item>
            </Chips>
            <Chips design='ghost'>
                <Chips.Item id='1' tabIndex={1} active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' tabIndex={1}>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' tabIndex={1}>
                    Chip
                </Chips.Item>
            </Chips>
        </>
    );
};
export const AllSizes: Story<ChipsProps> = (args) => {
    const [active, setActive] = React.useState<string | null>(null);
    const onClick = (event) => setActive(event.target.id);
    return (
        <Chips onClick={onClick} active={active}>
            <Chips.Item size='xs' id='1' active>
                Chip
            </Chips.Item>
            <Chips.Item size='s' id='2'>
                Chip
            </Chips.Item>
            <Chips.Item size='m' id='3'>
                Chip
            </Chips.Item>
            <Chips.Item size='l' id='4'>
                Chip
            </Chips.Item>
            <Chips.Item size='xl' id='5'>
                Chip
            </Chips.Item>
            <Chips.Item size='xxl' id='6'>
                Chip
            </Chips.Item>
        </Chips>
    );
};
export const RightAligned: Story<ChipsProps> = (args) => {
    return (
        <Chips align='right'>
            <Chips.Item id='1' active>
                Chip
            </Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );
};
export const DisabledChips: Story<ChipsProps> = (args) => {
    return (
        <>
            <Chips design='accent'>
                <Chips.Item id='1' disabled active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' disabled>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' disabled>
                    Chip
                </Chips.Item>
            </Chips>
            <Chips design='primary'>
                <Chips.Item id='1' disabled active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' disabled>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' disabled>
                    Chip
                </Chips.Item>
            </Chips>
            <Chips design='ghost'>
                <Chips.Item id='1' disabled active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' disabled>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' disabled>
                    Chip
                </Chips.Item>
            </Chips>
        </>
    );
};

export const ChipsWithOnClickOnItems: Story<ChipsProps> = (args) => {
    const [active, setActive] = React.useState('1');
    return (
        <>
            <Chips active={active}>
                <Chips.Item id='1' onClick={() => setActive('1')}>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' disabled onClick={() => setActive('2')}>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' onClick={() => setActive('3')}>
                    Chip
                </Chips.Item>
            </Chips>
        </>
    );
};

export const ViewOnly: Story<ChipsProps> = (args) => {
    const [active, setActive] = React.useState(['1', '2']);

    return (
        <Chips active={active} viewOnly>
            <Chips.Item size='l' id='1'>
                Chip 1
            </Chips.Item>
            <Chips.Item size='l' id='2'>
                Chip 2
            </Chips.Item>
            <Chips.Item size='l' id='3'>
                Chip 3
            </Chips.Item>
            <Chips.Item size='l' id='4'>
                Chip 3
            </Chips.Item>
        </Chips>
    );
};
