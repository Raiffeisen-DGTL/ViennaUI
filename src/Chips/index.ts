export { Chips, type ChipsProps, type ChipsItemProps } from './Chips';
