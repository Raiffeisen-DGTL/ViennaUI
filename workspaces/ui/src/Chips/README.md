# Chips

Компонент `Chips` используется для сегментации и фильтрации контента.


## Импорт

```
import { Chips } from 'vienna-ui';
```

## Свойства / Props

Prop | Type | Default | Description
--- | --- | --- | ---
active | string \| string[] \| undefined | false | Активные элементы
design | 'accent' \| 'primary' \| 'ghost' \| undefined | 'primary' |
align | string \| undefined | 'left' |
onClick | (event: React.FormEvent) => void | false |

## Свойства элемента / Item Props

Prop | Type | Default | Description
--- | --- | --- | ---
id | string \| undefined | false | Идентификатор элемента
active | boolean \| undefined | false | Активный элемент
disabled | boolean \| undefined | false | Заблокированный элемент
onFocus | ((event: FormEvent) => void) \| undefined | false |
onBlur | ((event: FormEvent) => void) \| undefined | false |
onKeyUp | ((event: FormEvent) => void) \| undefined | false |
onKeyDown | ((event: FormEvent) => void) \| undefined | false |
onMouseUp | ((event: FormEvent) => void) \| undefined | false |
onMouseDown | ((event: FormEvent) => void) \| undefined | false |
onMouseEnter | ((event: FormEvent) => void) \| undefined | false |
onMouseOut | ((event: FormEvent) => void) \| undefined | false |
onMouseLeave | ((event: FormEvent) => void) \| undefined | false |

## Использование

Компонент состоит из родительского контейнера `Chips` и дочерних элементов `Chips.Item`.

> Компонент является контролируемым, то есть чтобы отобразить элементы как активные,
> необходимо получить их значение через обработчик `onChange` и прокинуть в `active`.

```jsx
<Chips>
    <Chips.Item id='1' active>
        Chip
    </Chips.Item>
    <Chips.Item id='2'>Chip</Chips.Item>
    <Chips.Item id='3'>Chip</Chips.Item>
</Chips>
```

## Дизайн
##### Свойство `design`

Комопнент поддерживает 3 опции внешнего вида: `accent`, `primary` и `ghost`

```jsx
<Chips design='accent'>
    <Chips.Item id='1' tabIndex='1' active>
        Chip
    </Chips.Item>
    <Chips.Item id='2' tabIndex='1'>
        Chip
    </Chips.Item>
    <Chips.Item id='3' tabIndex='1'>
        Chip
    </Chips.Item>
</Chips>
```

## Выравнивание
##### Свойство `align`

```jsx
<Chips align='right'>
    <Chips.Item id='1' active>
        Chip
    </Chips.Item>
    <Chips.Item id='2'>Chip</Chips.Item>
    <Chips.Item id='3'>Chip</Chips.Item>
</Chips>
```

## Активный элемент
##### Свойство `active`

Активный элемент мможно выбрать, как установив свойство `active` для компонента `Chips`, который принимает строку или массив строк `id` компонентов `Chips.Item`,
так и установив свойство `active` для компонента `Chips.Item` в `true`.

```jsx
{() => {
    const [active, setActive] = React.useState(null);
    const onClick = (event) => setActive(event.target.id);
    return (
        <Chips design='primary' onClick={onClick} active={active}>
            <Chips.Item id='1' tabIndex='1'>
                Chip
            </Chips.Item>
            <Chips.Item id='2' tabIndex='1'>
                Chip
            </Chips.Item>
            <Chips.Item id='3' tabIndex='1'>
                Chip
            </Chips.Item>
            <Chips.Item id='4' tabIndex='1'>
                Chip
            </Chips.Item>
        </Chips>
    );
}}
```

```jsx
<Chips design='primary' onClick={onClick}>
    <Chips.Item id='1' tabIndex='1' active>
        Chip
    </Chips.Item>
    <Chips.Item id='2' tabIndex='1'>
        Chip
    </Chips.Item>
    <Chips.Item id='3' tabIndex='1'>
        Chip
    </Chips.Item>
    <Chips.Item id='4' tabIndex='1'>
        Chip
    </Chips.Item>
</Chips>
```

## Заблокированный элемент
##### Свойство `disabled`

```jsx
<Chips design='accent'>
    <Chips.Item id='1' disabled active>
        Chip
    </Chips.Item>
    <Chips.Item id='2' disabled>
        Chip
    </Chips.Item>
    <Chips.Item id='3' disabled>
        Chip
    </Chips.Item>
</Chips>
```
