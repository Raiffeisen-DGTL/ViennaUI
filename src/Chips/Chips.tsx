import React, {
    FC,
    Children,
    isValidElement,
    KeyboardEventHandler,
    useCallback,
    useRef,
    useEffect,
    ReactNode,
    ReactElement,
} from 'react';
import { Box, PropsBox, BoxItem, PropsBoxItem, ChipsViewOnlyText } from './Chips.styles';
import { withTabFocusState, reactNodeIsComponent, SizeType } from '../Utils';
import { BoxStyled } from '../Utils/styled';
import { omit } from '../Utils/omit';
import { ViewOnly, WithViewOnly } from '@/ViewOnly';

export interface BaseItemProps {
    active?: boolean;
    size?: SizeType<'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl'>;
    disabled?: boolean;
}

export interface ChipsProps extends BoxStyled<typeof Box, PropsBox>, WithViewOnly {
    size?: PropsBox['$size'];
    design?: PropsBox['$design'];
    align?: PropsBox['$align'];
    active?: null | string | string[];
    onPressEnter?: KeyboardEventHandler<HTMLDivElement>;
}
export interface ChipsItemProps extends Omit<BoxStyled<typeof BoxItem, PropsBoxItem>, 'onClick'> {
    size?: PropsBoxItem['$size'];
    design?: PropsBoxItem['$design'];
    active?: PropsBoxItem['$active'];
    disabled?: PropsBoxItem['$disabled'];
    isFocusStateVisible: boolean;
    onClick?: (e: MouseEvent | React.MouseEvent<HTMLDivElement | HTMLSpanElement>) => void;
}

const isActive = ({ active, id }: ChipsItemProps, state?: null | string | string[]) => {
    if (!state) {
        return active;
    }

    return Array.isArray(state) && id ? state.includes(id) : state === id;
};

const ItemWithEvent = React.forwardRef<HTMLDivElement, ChipsItemProps>(
    ({ onClick, size = 's', design, active, disabled, isFocusStateVisible, ...attrs }, ref) => {
        const item = useRef<HTMLDivElement>(null);

        useEffect(() => {
            const clickHandler = (e: MouseEvent) => {
                onClick?.(e);
            };

            if (item.current) {
                item.current.addEventListener('click', clickHandler, true);
            }

            return () => item.current?.removeEventListener('click', clickHandler, true);
        }, [onClick]);

        return (
            <BoxItem
                {...attrs}
                ref={ref || item} // Используем переданный ref или внутренний ref
                as='div'
                data-checked={active ? 'true' : 'false'}
                $size={size}
                $design={design}
                $active={active}
                $disabled={disabled}
                $isFocusStateVisible={isFocusStateVisible}
            />
        );
    }
);

const ChipItem = withTabFocusState(ItemWithEvent);

const Item = (props: Omit<ChipsItemProps, 'isFocusStateVisible'>) => {
    const newProps = { ...props };
    delete newProps.active;
    delete newProps.size;
    delete newProps.disabled;
    delete newProps.design;
    return <span {...(newProps as Omit<typeof newProps, 'active' | 'size' | 'disabled' | 'design'>)} />;
};

export const Chips: FC<ChipsProps> & { Item: typeof Item } = ({
    design = 'primary',
    align = 'left',
    size = 's',
    children,
    active: rootActive,
    viewOnly,
    viewOnlyText,
    onPressEnter,
    onClick,
    ...attrs
}): JSX.Element => {
    const handleKeyDown = useCallback<KeyboardEventHandler<HTMLDivElement>>(
        (event) => {
            switch (event.key) {
                case ' ':
                case 'Enter':
                    event.preventDefault();
                    onPressEnter?.(event);
            }

            return null;
        },
        [onPressEnter]
    );

    const renderChipItems = (child: ReactNode | ReactNode[]): ReactNode => {
        if (Array.isArray(child)) {
            return child.map((item, index) => {
                return <React.Fragment key={index}>{renderChipItem(item)}</React.Fragment>;
            });
        }

        return renderChipItem(child);
    };

    const renderChipItem = (child: ReactNode): ReactNode => {
        if (reactNodeIsComponent(child, Item)) {
            const { props: chipProps } = child as ReactElement<ChipsItemProps>;
            const onKeyDown = chipProps.onKeyDown || handleKeyDown;
            const active = isActive(chipProps, rootActive);
            const tabIndex = chipProps?.disabled ? -1 : chipProps.tabIndex;
            return (
                <ChipItem
                    role='option'
                    active={active}
                    design={design}
                    size={chipProps.size || size}
                    onKeyDown={onKeyDown}
                    tabIndex={tabIndex}
                    onClick={
                        !chipProps.disabled ? ((chipProps.onClick || onClick) as ChipsItemProps['onClick']) : undefined
                    }
                    {...(omit(chipProps, 'tabIndex', 'onClick', 'onKeyDown') as Omit<
                        ChipsItemProps,
                        'ref' | 'tabIndex' | 'onClick' | 'onKeyDown'
                    >)}
                />
            );
        }
        if (isValidElement(child) && (child as React.ReactElement<{ children: React.ReactNode }>).props.children) {
            const childProps = (child as React.ReactElement<{ children: React.ReactNode }>).props;
            return React.cloneElement(child, childProps, renderChipItems(childProps.children));
        }

        return child;
    };

    const content = Children.map(children, renderChipItems);

    if (viewOnly) {
        const chipItemsProps: ChipsItemProps[] = [];
        const getChipItemsProps = (child: ReactNode | ReactNode[]) => {
            if (Array.isArray(child)) {
                child.forEach(getChipItemsProps);
            } else if (reactNodeIsComponent(child, Item)) {
                chipItemsProps.push(child.props as ChipsItemProps);
            } else if (
                isValidElement(child) &&
                (child as React.ReactElement<{ children: React.ReactNode }>).props.children
            ) {
                getChipItemsProps(child.props.children);
            }
        };
        Children.forEach(children, getChipItemsProps);
        const activeChips = chipItemsProps
            .filter((chipProps) => isActive(chipProps, rootActive))
            .map((item) => item.children);
        return (
            <ViewOnly>
                <ChipsViewOnlyText $size={size}>
                    {viewOnlyText ??
                        activeChips.map((item, index, arr) => (
                            <React.Fragment key={index}>
                                {item}
                                {index + 1 < arr.length && ', '}
                            </React.Fragment>
                        ))}
                </ChipsViewOnlyText>
            </ViewOnly>
        );
    }

    return (
        <Box {...attrs} $design={design} $align={align} $size={size}>
            {content}
        </Box>
    );
};

Chips.displayName = 'Chips';

Chips.Item = Item;
Item.displayName = 'Item';
