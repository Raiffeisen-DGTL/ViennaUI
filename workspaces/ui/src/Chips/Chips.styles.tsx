import styled, { css } from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';
import { ChipsProps, ItemProps } from './Chips';

const presets = getPresets('chips', {
    box: null,
    base: null,
    focus: null,
    design: 'design',
    disabled: 'design',
    hover: null,
    custom: null,
    boxCustom: null,
});

const active = getPresets('chips.active', {
    design: 'design',
    hover: 'design',
    disabled: 'design',
    custom: null,
});

const justify = (props) => (props.align === 'right' ? 'flex-end' : 'flex-start');

export const Box = styled.div<ChipsProps>`
    display: flex;
    box-sizing: border-box;
    position: relative;
    align-items: center;
    justify-content: ${justify};

    ${presets.box}
    ${presets.boxCustom}
`;

export const Item = styled.div<ItemProps>`
    display: inline-flex;
    justify-content: center;
    align-items: center;
    box-sizing: border-box;

    ${presets.base}
    ${presets.design}

    &:hover {
        ${presets.hover}
    }

    ${({ disabled }) =>
        disabled &&
        css`
            pointer-events: none;
            ${presets.disabled};
        `}

    ${(props) =>
        props.active &&
        css`
            ${active.design}

            &:hover {
                ${active.hover}
            }

            ${props.disabled &&
            css`
                ${active.disabled};
            `}

            ${active.custom}
        `}

    ${({ isFocusStateVisible }: any) =>
        isFocusStateVisible &&
        css`
            &:focus {
                ${presets.focus}
            }
        `}

    ${presets.custom}
`;
