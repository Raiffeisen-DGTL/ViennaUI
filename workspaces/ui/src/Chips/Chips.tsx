import React, { useCallback } from 'react';
import { Box, Item } from './Chips.styles';
import { withTabFocusState } from '../Utils';

interface Props {
    children?: any;
    active?: string | string[];
    design?: 'accent' | 'primary' | 'ghost';
    align?: string;
    onClick?: (event: React.FormEvent) => void;
}

interface HTMLAttributeProps {
    [key: string]: any;
    id?: string;
    title?: string;
    className?: string;
    role?: string;
    spellCheck?: boolean;
    tabIndex?: number;
}

export interface ItemProps {
    active?: boolean;
    disabled?: boolean;
    onFocus?: (event: React.FormEvent) => void;
    onBlur?: (event: React.FormEvent) => void;
    onKeyUp?: (event: React.FormEvent) => void;
    onKeyDown?: (event: React.FormEvent) => void;
    onMouseUp?: (event: React.FormEvent) => void;
    onMouseDown?: (event: React.FormEvent) => void;
    onMouseEnter?: (event: React.FormEvent) => void;
    onMouseOut?: (event: React.FormEvent) => void;
    onMouseLeave?: (event: React.FormEvent) => void;
}

export type ChipsProps = Props & HTMLAttributeProps;
export type ChipsItemProps = ItemProps & HTMLAttributeProps;

const isActive = (props: any, state?: string | string[]) => {
    if (!state) {
        return props.active;
    }

    return Array.isArray(state) ? state.includes(props.id) : state === props.id;
};

const ChipItem = withTabFocusState(Item);

export const Chips: React.FC<ChipsProps> & { Item: React.FC<ChipsItemProps> } = (props): JSX.Element => {
    const { design, onClick, children, active: setActive, ...attrs } = props;

    const handleKeyDown = useCallback(
        (event: React.KeyboardEvent) => {
            switch (event.key) {
                case ' ':
                case 'Enter':
                    event.preventDefault();
                    return onClick?.(event);
            }

            return null;
        },
        [onClick]
    );

    const content = React.Children.map(React.Children.toArray(children), (child) => {
        if (React.isValidElement(child)) {
            const { props: chipProps } = child;
            const onKeyDown = chipProps.onKeyDown || handleKeyDown;

            const active = isActive(chipProps, setActive);
            const tabIndex = chipProps.disabled ? '-1' : chipProps.tabIndex;
            const newProps = { ...chipProps, ...{ active, design, onClick, onKeyDown, tabIndex } };

            return <ChipItem role='option' {...newProps} />;
        }

        return null;
    });

    return (
        <Box design={design} {...attrs}>
            {content}
        </Box>
    );
};

Chips.displayName = 'Chips';
Chips.defaultProps = {
    design: 'primary',
    align: 'left',
};

Chips.Item = (props: ChipsItemProps): JSX.Element => {
    return <span {...props} />;
};

export default Chips;
