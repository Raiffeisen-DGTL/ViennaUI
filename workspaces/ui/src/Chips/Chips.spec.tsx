import React from 'react';
import { Chips, ChipsProps } from './Chips';

const designs: ChipsProps['design'][] = ['accent', 'primary', 'ghost'];

test('Chips', () => {
    const snap = snapshot.render(
        <Chips>
            <Chips.Item id='1' active>
                Chip
            </Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );
    expect(snap).toMatchSnapshot();
});

test('Chips w/ design', () => {
    designs.forEach((design) => {
        const snap = snapshot.render(
            <Chips design={design}>
                <Chips.Item id='1' active>
                    Chip
                </Chips.Item>
                <Chips.Item id='2'>Chip</Chips.Item>
                <Chips.Item id='3'>Chip</Chips.Item>
            </Chips>
        );
        expect(snap).toMatchSnapshot();
    });
});

test('Chips w/ align', () => {
    const snap = snapshot.render(
        <Chips align='right'>
            <Chips.Item id='1' active>
                Chip
            </Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );
    expect(snap).toMatchSnapshot();
});

test('Chips w/ active', () => {
    const snap = snapshot.render(
        <Chips active='1'>
            <Chips.Item id='1'>Chip</Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );
    expect(snap).toMatchSnapshot();
});

test('Chips w/ multiple active', () => {
    const snap = snapshot.render(
        <Chips active={['2', '3']}>
            <Chips.Item id='1'>Chip</Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );
    expect(snap).toMatchSnapshot();
});

test('Chips w/ disabled', () => {
    designs.forEach((design) => {
        const snap = snapshot.render(
            <Chips design={design}>
                <Chips.Item id='1' active disabled>
                    Chip
                </Chips.Item>
                <Chips.Item id='2' disabled>
                    Chip
                </Chips.Item>
                <Chips.Item id='3' disabled>
                    Chip
                </Chips.Item>
            </Chips>
        );
        expect(snap).toMatchSnapshot();
    });
});

test('Chips w/ any props', () => {
    const props = { className: 'className', id: 'id' };
    const component = (
        <Chips {...props}>
            <Chips.Item id='1'>Chip</Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();

    const wrapper = snapshot.shallow(component);
    for (const [key, value] of Object.entries(props)) {
        expect(wrapper.props()[key]).toEqual(value);
    }
});

test('Chips w/ onClick', () => {
    const fn = jest.fn();

    const component = (
        <Chips onClick={fn}>
            <Chips.Item id='1'>Chip</Chips.Item>
            <Chips.Item id='2'>Chip</Chips.Item>
            <Chips.Item id='3'>Chip</Chips.Item>
        </Chips>
    );

    const wrapper = snapshot.shallow(component);
    const chip = wrapper.childAt(0);

    chip.simulate('click');
    chip.simulate('keydown', { key: 'Enter', preventDefault: () => null });
    chip.simulate('keydown', { key: ' ', preventDefault: () => null });

    expect(fn).toHaveBeenCalledTimes(3);
});
