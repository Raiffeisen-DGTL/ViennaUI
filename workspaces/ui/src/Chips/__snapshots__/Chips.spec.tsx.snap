// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 fymOnN"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ active 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 fymOnN"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ align 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 yPYKc"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 fymOnN"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ any props 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr className"
  id="id"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ design 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 GjXIy"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ design 2`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 fymOnN"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ design 3`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 hcRtct"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kohHQH"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kohHQH"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ disabled 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 jAAPqS"
    disabled=""
    id="1"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 iCxqMm"
    disabled=""
    id="2"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 iCxqMm"
    disabled=""
    id="3"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ disabled 2`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 lhvWtl"
    disabled=""
    id="1"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 iCxqMm"
    disabled=""
    id="2"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 iCxqMm"
    disabled=""
    id="3"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ disabled 3`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kGYMjp"
    disabled=""
    id="1"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 jOYZan"
    disabled=""
    id="2"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 jOYZan"
    disabled=""
    id="3"
    role="option"
    tabindex="-1"
  >
    Chip
  </div>
</div>
`;

exports[`Chips w/ multiple active 1`] = `
<div
  class="Chipsstyles__Box-sc-1xafpmg-0 jnPUVr"
>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 kOgGBY"
    id="1"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 fymOnN"
    id="2"
    role="option"
  >
    Chip
  </div>
  <div
    class="Chipsstyles__Item-sc-1xafpmg-1 fymOnN"
    id="3"
    role="option"
  >
    Chip
  </div>
</div>
`;
